gematria/datasets/pipelines/compile_modules_lib.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2024 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import apache_beam as beam
+from rules_python.python.runfiles import runfiles
+import os
+
+from absl import logging
+
+from collections.abc import Iterable
+from collections.abc import Callable
+
+import subprocess
+
+from gematria.datasets.python import extract_bbs_from_obj
+
+
+def _get_llvm_binary_path(tool_name: str) -> str:
+  runfiles_env = runfiles.Create(os.environ)
+  assert runfiles_env is not None
+  return runfiles_env.Rlocation('llvm-project/llvm/' + tool_name)
+
+
+class OptimizeModules(beam.DoFn):
+  """A Beam function that uses LLVM opt to optimize bitcode modules."""
+
+  def __init__(self, optimization_pass_lists: list[str]):
+    self.optimization_pass_lists = optimization_pass_lists
+    self.opt_path = _get_llvm_binary_path('opt')
+
+  def optimize_module(
+      self, input_module: bytes, optimization_pass_list: list[str]
+  ) -> bytes:
+    command_vector = [self.opt_path, f'-passes={optimization_pass_list}']
+    with subprocess.Popen(
+        command_vector,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    ) as opt_process:
+      (output_bc, stderr) = opt_process.communicate(input=input_module)
+      if opt_process.returncode != 0:
+        logging.error(stderr)
+        raise ValueError('Expected opt to return 0')
+      return output_bc
+
+  def process(self, input_module: bytes) -> Iterable[bytes]:
+    for optimization_pass_list in self.optimization_pass_lists:
+      try:
+        yield self.optimize_module(input_module, optimization_pass_list)
+      except ValueError:
+        continue
```

Comment:
It could be useful to use state/metrics to track the number of failed modules: https://beam.apache.org/documentation/programming-guide/#combiningstate

Same for the other fn's where we drop modules on error.

Reply:
Good point, but it doesn't seem like there's an easy to get the values out after the `DoFn` has been run?

In ordered to read the value of a `CombiningValueStateSpec`, it seems like you need a `DoFn.StateParam` as the parameter to the `process` function of a `DoFn`?

Reply:
I'll look into at some point. Seems like getting it working in a separate patch would be ideal.